src/nfa/nfa.cc

```diff
@@ -312,6 +312,136 @@ struct StackLevel {
     };
 };
 
+struct NodeData {
+    Post::const_iterator post_it;
+    Post::const_iterator post_end;
+    StateSet::const_iterator targets_it{};
+    StateSet::const_iterator targets_end{};
+    unsigned long index;
+    unsigned long lowlink;
+    bool initilized;
+    bool on_stack;
+
+    NodeData() = default;
+
+    NodeData(State q, const Delta & delta, unsigned long index) 
+        : post_it(delta[q].cbegin()), post_end(delta[q].cend()), index(index), lowlink(index), initilized(true), on_stack(true) {
+        if (post_it != post_end) {
+            targets_it = post_it->cbegin();
+            targets_end = post_it->cend();
+        }
+    };
+
+    // TODO: this sucks. In fact, if you want to check that you have the last sucessor, you need to 
+    // first align the iterators.
+    // TODO: this is super-ugly. If we introduce Post::transitions iterator, this could be much easier.
+    void align_succ() {
+        while (this->post_it != this->post_end && this->targets_it == this->targets_end) {
+            if (this->targets_it == this->targets_end) {
+                ++this->post_it;
+                if (this->post_it != this->post_end) {
+                    this->targets_it = this->post_it->cbegin();
+                    this->targets_end = this->post_it->cend();
+                }
+            } 
+        }
+    }
+
+    State get_act_succ() {
+        align_succ();
+        return *(this->targets_it);
+    }
+
+    void move_next_succ() {
+        if(this->post_it == this->post_end) {
+            return;
+        }
+        this->targets_it++;
```

Comment:
```suggestion
        ++this->targets_it;
```
Might be a teensy-weensy-bit faster.